### PR TITLE
Ensure external processes are killed when test process exits

### PR DESCRIPTION
**Pull request by mikeharder**
- Prevents locked files when stop debugging unit test
- Addresses #1279

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs [hunk 1]
```diff
@@ -254,6 +254,11 @@ private void WaitUntilSiteStarted()
                 }
 
                 HostProcess = Process.GetProcessById(workerProcess.ProcessId);
+
+                // Ensure w3wp.exe is killed if test process termination is non-graceful.
+                // Prevents locked files when stop debugging unit test.
+                ProcessTracker.Add(HostProcess);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: How can this work for w3wp? The worker process would be immediately recycled and the dll would be locked again when killed.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/ProcessTracker.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
+using System;
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/ProcessTracker.cs]: nit copyright.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/ProcessTracker.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Server.IntegrationTesting.IIS
+{
+    // Uses Windows Job Objects to ensure external processes are killed if the current process is terminated non-gracefully.
+    internal static class ProcessTracker
+    {
+        private static readonly IntPtr _jobHandle;
+
+        static ProcessTracker()
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/ProcessTracker.cs]: Where did you get this class?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs (hunk 1 above)
Comment (mikeharder) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: When the test process exits, `w3wp.exe` (and the `dotnet.exe` child process of `w3wp.exe`) are killed.  It's true that `w3wp.exe` is immediately restarted, but the `dotnet.exe` child process will not.  So, while some dlls may still be locked, it seems better than doing nothing.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/ProcessTracker.cs (hunk 3 above)
Comment (mikeharder) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/ProcessTracker.cs]: Slightly modified version of https://stackoverflow.com/a/37034966/102052

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: The file that is locked is always aspnetcore[v2].dll, so this really doesn't help at all.

Comment (mikeharder) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: I would argue its still slightly better to kill `dotnet.exe` than do nothing.  However, I'm also fine limiting this change to IIS Express, if we think it's misleading or noisy to try killing `w3wp.exe`.

Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: Anyway to register a callback? Doesn't look like it, but maybe we can add a callback to undo the redirection.config changes on exit.

Comment (mikeharder) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: I believe the best we can do with Windows Job Objects is to kill processes.  If we wanted to undo the `redirection.config` changes (or anything else), we could create a watchdog process which detects when the test process terminates (e.g. via anonymous pipes) and then performs cleanup.

**Pull request merged**